### PR TITLE
Config file as argument or in env

### PR DESCRIPTION
**Pull request by teoti001**
A first implementation of a new way to handle the configuration of the caramel tools/scripts as requested in issue  #55. This first implementation makes it possible for the user to set the path to the .ini-file in the environment(the variable CARAMEL_INI) instead of as an argument to the script. The idea being to make it possible in the future to be able to configure via the setting of environment variables instead of a .ini-file, while also still being compatible with the current way of configuration.

### CODE REVIEW

File: caramel/config.py
Comment (Spindel) [inline caramel/config.py:19]: Check if this could be turned into an argparse action instead, last part of "action" documentation.

1: https://docs.python.org/3/library/argparse.html#action

File: caramel/config.py
Comment (Spindel) [inline caramel/config.py:1]: This file should not have a shebang as it should not be executable on it's own.

File: tests/fixtures.py
Comment (Spindel) [inline tests/fixtures.py:20]: datetime import should be  high up as it's an python standardlib import.  

import order should be:

stdlib  then   exeternal modules last  internal modules

File: tests/test_config.py
Comment (Spindel) [inline tests/test_config.py:51]: No need to use a fixture here for a single test-case, that only obscures what the test-case does.

File: tests/test_config.py
Comment (Spindel) [inline tests/test_config.py:26]: This permanently changes the environment without restoring it after the test is done, thus if you run the tests in arbitary order, this change will make the  other tests behave differently.

The solution to that is usually to make  a function liek "add_inifile_argument"  take an incoming environment dict as it's argument, with the default being "os.environ" that can be overridden for the tests.

**Review by Spindel — CHANGES_REQUESTED** [review]

File: caramel/config.py
Comment (teoti001) [reply caramel/config.py:1]: All .py files in the project have them, so was only following the precedent.

File: tests/test_config.py
Comment (teoti001) [reply tests/test_config.py:26]: My understanding was that changes made through os on the environment was not persistent? But I will defer to your judgement.

File: tests/test_config.py
Comment (teoti001) [reply tests/test_config.py:51]: I understand that, but I've locally continued development on adding loggers and their test uses an extended version of this fixture(in my sketches), so is there really a need to change this here when it will be extended and used in the near future(<1 month)?

File: tests/test_config.py
Comment (Spindel) [reply tests/test_config.py:26]: Persistent is a different thing.  It may not persist after you exit the process, but it's certainly a global state for the current process runtime, and other tests that run later will definitely see changes.

A simple demonstration:

```python
#!/bin/env python3
import os
import unittest

class TestOne(unittest.TestCase):
    def test_one(self):
        assert "HUMBUG" not in os.environ
        os.environ["HUMBUG"] = "Bah. Humbug"

    def test_two(self):
        assert "HUMBUG" not in os.environ
        os.environ["HUMBUG"] = "Bah. Humbug"

    def test_three(self):
        assert "HUMBUG" not in os.environ
        os.environ["HUMBUG"] = "Bah. Humbug"


if __name__ == "__main__":
    unittest.main()
```

File: tests/test_config.py
Comment (Spindel) [reply tests/test_config.py:51]: Premature abstractions and abstractions with a single user are generally to be avoided, especially when spread out across different files.

Overall, what you want to optimize for is making the single case easy to understand. You've already hit the impact of abstract difficulty there was to work with the fixtures in the model test-cases, and those are shared between  view and model tests.

So, since this (config parsing) doesn't involve any shared state, the fixture and abstractions should be as close as possible to the tests, in this case, preferrably in the same case.

Just as you currently have no use for inheriting "SImilarlyComparable" as the tests don't actually use that.

Comment (teoti001) [reply tests/test_config.py:51]: Nvm, Action-based implementation of the check for an ini-file made it redundant in this patch.

File: caramel/config.py
Comment (teoti001) [reply caramel/config.py]: It could indeed and it feels like a better solution.

File: tests/test_config.py
Comment (Spindel) [inline tests/test_config.py:19]: No returns needed from test-cases

```suggestion
```

File: tests/test_config.py
Comment (Spindel) [inline tests/test_config.py:30]: No returns needed from test-cases

```suggestion
```

File: tests/test_config.py
Comment (Spindel) [inline tests/test_config.py:15]:
```suggestion
        my_ini = "/a/path/to/my_ini_file.ini"
        parser = argparse.ArgumentParser()
        config.add_inifile_argument(parser, {})
```

File: caramel/scripts/tool.py
Comment (Spindel) [inline caramel/scripts/tool.py:71]: unrelated whitespace change for this commit

File: caramel/scripts/autosign.py
Comment (Spindel) [inline caramel/scripts/autosign.py:26]: unrelated whitespace change for this commit

**Review by Spindel — APPROVED** [review]
Final change looks good. Commit messages could use some work, mix of tenses etc. ( I prefer imperative, present tense, where the reader can interprete the commit message as a command to the code, and the diff as the result of performing that command.)  

I have some opinions about bundling changes from various tools together into the same commit, as well as running tools on files that otherwise have no changes in them, but since this project doesn't run in our CI, I'm not going to block on that.

Comments on a few of the commits which introduces spurious changes to other parts of the files.

**Comment by Spindel** [comment]
okay, since this has exploded, I'm going to request that you make a separate MR for the black + lint fixes in order to make it easier to review the actual changes separately.

File: tests/__init__.py
Comment (Spindel) [inline tests/__init__.py:12]: This is a functional change that may not be intended.   it assumes "tests" is an installed / available python package rather than being a local import.

File: tests/fixtures.py
Comment (Spindel) [inline tests/fixtures.py:126]: These are not test fixtures. They are validation or test-matrices.

File: tests/test_config.py
Comment (Spindel) [inline tests/test_config.py:27]: Split inifile tests and verbosity tests. They are different test-cases.

File: caramel/config.py
Comment (Spindel) [inline caramel/config.py:17]: Why is the shared-common behaviour for all code isolated and make the other code adapt to it?
The pattern of changes introduced here is:

1) "get an argument parser"
2) Add our own argument parsers
3) Extend it with the common defaults
4) parse our own changes
5) perform some common default operatoion

When the code could simply be a:

parser = get_common_parser(....)

Where the common parser would add "verbosity" and "ini file"  settings as "reasonable" defaults.

See exaple below.

File: caramel/scripts/autosign.py
Comment (Spindel) [inline caramel/scripts/autosign.py:85]:
```
parser = config.get_argparser()
parser.add_argument("--delay", ..)
parser.add_argument("--valid", ..)
```

**Review by Spindel — CHANGES_REQUESTED** [review]
I was about to merge the INI file changes as-is, and didn't expect you to continue adding features to that MR after I had approved it, anyhow, now it's sprawling, adding functional changes mixed with tooling changes across the tree.  I'd appreciate it if you split out the lint/whitespace changes to a separate branch that can stand on it's own.

**Comment by teoti001** [comment]
Reset this branch 3 commits

File: tests/__init__.py
Comment (teoti001) [reply tests/__init__.py:12]: I cannot for the life of me remember exactly what it was, but a linter or debugger got caught up on it, and since the "from 'local package' import module" semantic is used in multiple cases of the code thats what i substituted it for.

File: tests/fixtures.py
Comment (teoti001) [reply tests/fixtures.py]: Indeed, but I was weighing adding additional files, which wouldnt necessarily help with readability and since the naming convention is [...]Data, I found it logically to add additional data there.

File: caramel/config.py
Comment (teoti001) [reply caramel/config.py]: I disagree, "simple" monolithic implementations can sometimes just be obfuscating I think. In this case the suggestion would compile 3 lines of code to 1 in every tool, but will obfuscate what happens/make it necessary to go into get_common_parser to see what "common" parser options are. I think keeping it separated and modular with descriptive function names will make it easier to understand the code on a first overview.

**Comment by teoti001** [comment]
Waiting for #60 #61, and will rebase this branch after that.

File: caramel/config.py
Comment (teoti001) [reply caramel/config.py]: see pull req #62

**Pull request merged**
